Introduction
============

VAS is software to aggregate variant information sourced from multiple VCF files (and some others, see below) into a variety of summary files with varying levels of detail and statistics. The outputs range from high-level summaries to full, detailed reports in text and Microsoft Excel formats. The intended audience is both biologists and bioinformaticians.

System Requirements
===================

Python modules
--------------

The following modules are required:

* numpy (http://www.numpy.org/)
* pandas (http://pandas.pydata.org/)
* HTSeq (http://www-huber.embl.de/HTSeq/)

The following modules are optional:

* pytabix (https://github.com/slowkow/pytabix)
* xlwt

Additional tools
----------------

* Tabix (optional, see _Annotation_, below; https://github.com/samtools/tabix)

Annotation
----------

VAS requires a reference annotation file (in GTF or GFF3 format) for feature definition. Contig (chromosome) names must match between the input VCF and reference annotation file; for example, for _Homo sapiens_, UCSC uses the form **chr17** while Ensembl! uses the form **17**. Protip: If your human data are aligned and variants called with _chr_ style contig names but you wish to use Ensembl genes, you can try the GENCODE annotation which contains most Ensembl genes but using _chr_ style contig numbering.

Databases
---------

VAS can _optionally_ check variants against any number of supplied databases and report presence (including identifier) or absence in the database. Common choices would be one or more releases of dbSNP (e.g., dbSNP 137 and latest dbSNP; dbSNP clinvar and dbSNP all), 1000 Genomes, NHLBI 6500 Exomes, or COSMIC. These databses must be in VCF format, bgzipped, and indexed with tabix (https://github.com/samtools/tabix). 

Users may supply their own custom databases, provided it conforms to established VCF standard, is compressed with bgzip, and indexed with tabix.

Variant data
------------

These data are your primary source data that you wish VAS to aggregate and summarize. In theory VAS will work with any well-formed generic VCF, but it has been tested with and contains specific code to handle VCF files generated by the following tools:

* Ion Torrent PGM (default machine output)
* Illumina Miseq (default machine output)
* GATK
* GATK SomaticIndelDetector
* GATK HaplotypeCaller
* MuTect
* VarScan
* FreeBayes
* Samtools

In addition, VAS can read the more detailed .out files produced by MuTect.

Installation
============

git clone https://github.com/blachlylab/mucor.git

Setup and Operation
===================

Overview
--------
Running VAS to aggregate and summarize variants is a two step process: 1. project configuration/setup 2. aggregation

Project setup
-------------

Files and Operation
-------------------
User Interface Scripts:
mucor_config.py - Configure the run by creating a JSON file to pass to mucor.py
mucor.py - The main mucor script that parses data and generates output

Mucor Functions and Modules:
config.py - Contains the Config class, used in mucor.py to assign the JSON file to an object
databases.py - Contains a class and function used in mucor.py o check for known, annotated variants
inputs.py - Contains the Parser class, used in mucor.py to read variant call files
mucorfeature.py - Contains the MucorFeature class, used in mucor.py to store feature data
output.py - Contains the Writer class, used in mucor.py to write various output files
variant.py - Contains the Variant class, used in mucor.py to store data about each variant

Operation
-----------------------------
Operating mucor occurs in two distinct steps with their own python scripts; configuration and execution. 

Configuration: mucor_config.py
The configuration step is completed using the provided `mucor_config.py` utility. It accepts the following command line arguments and creates a JSON file that will be passed to the main Mucor script. All required parameters must be supplied. Optional parameters may be left blank, some of which will assume the given default values. 

  -ex, --example        Print a valid, example JSON config file and exit.
Function that will write a template of the JSON config. It can be edited manually and supplied to mucor. 

  -g GFF, --gff GFF
Reference annotation GFF/GTF for feature binning. Required

  -f FEATURETYPE, --featuretype FEATURETYPE 
Feature type into which to bin. Gencode GTF example: gene_name, gene_id, transcript_name, transcript_id, etc. Required

  -db DATABASES, --databases DATABASES
Colon delimited name and path to variant database in bgzipped VCF format. Can be declared >= 0 times. Ex: -db name1:/full/user/path/name1.vcf.gz. Optional

  -s SAMPLES, --samples SAMPLES 
Text file containing sample names. One sample per line. Configuration may not be correct if any sample names are within another sample name. Ex: U-23 and U-238. U-23 would erroneously identify U-238 files, requiring manual modification of the JSON file. Required

  -d PROJECT_DIRECTORY, --project_directory PROJECT_DIRECTORY 
Working/project directory, in which to find variant call files to aggregate. Variant calls can be in the provided directory, or any of its subdirectories. Default: current working directory

  -vcff VCF_FILTERS, --vcf_filters VCF_FILTERS 
Comma separated list of VCF filters to allow. Default: PASS

 -a ARCHIVE_DIRECTORY, --archive_directory ARCHIVE_DIRECTORY
Specify directory in which to read/write archived annotations. This step will significantly speed up future runs that use the same annotation and feature type, even if the sample data changes. Undefined will prevent using the annotation archive features. Optional

  -r REGIONS, --regions REGIONS 
Comma separated list of bed regions and/or bed files by which to limit output. Bed regions can be specific positions, or entire chromosomes. Ex: chr1:10230-10240,chr2,my_regions.bed. Optional

  -jco JSON_CONFIG_OUTPUT, --json_config_output JSON_CONFIG_OUTPUT
Name of JSON configuration output file. This is the configuration file fed into Mucor. Required

  -outd OUTPUT_DIRECTORY, --output_directory OUTPUT_DIRECTORY
Name of directory in which to write Mucor output. Required

  -outt OUTPUT_TYPE, --output_type OUTPUT_TYPE
Comma separated list of desired output types. Options include: counts, txt, longtxt, xls, longxls, bed, featXsamp, featmutXsamp, all. Default: counts,txt

Execution: mucor.py
Mucor is executed by running the main, mucor.py python script with the desired json configuration file as the only argument. This facilitates rerunning analyses in the future and documenting the selected parameters to maintain consistency between runs. 

python mucor.py configuration.json

Outputs
-----------------------------
Users may select any number of output formats from the list below. These must be specified in the JSON configuration file prior to executing the main Mucor script.  

counts
Print counts of mutations per feature. Output: counts.txt

txt
Print all information about each mutation, combining all instances of a mutation (irrespective of in how many samples it appears) into a single row. Useful for mutation-centric output for a project. Identical to xls in layout, but different in file format. Output: variant_details.txt

longtxt
Similar to txt above, but writes each instance of a mutation to a new row. Each mutation is written once per source file, instead of combining reoccurring mutations in to 1 unique row. Identical to longxls in layout, but different in file format. Output: long_variant_details.txt

xls
Print all information about each mutation, combining all instances of a mutation (irrespective of in how many samples it appears) into a single row. Useful for mutation-centric output for a project. Identical to txt in layout, but different in file format. Output: variant_details.xls

longxls
Similar to xls above, but writes each instance of a mutation to a new row. Each mutation is written once per source instead of combining reoccurring mutations in to 1 unique row. Identical to longtxt in layout, but different in file format. Output: long_variant_details.xls

bed
Print bed file of the variant locations Output: variant_locations.bed

vcf
Print vcf file of the variant locations, features, depths, and variant frequencies. Output: variant_locations.vcf

featXsamp
Print table of mutation counts per feature per sample. Sample names populate the table header and feature names populate the first column. The count of unique mutations per sample per feature are the table values. Output: feature_by_sample.txt

featmutXsamp
Print table of mutations per sample. Sample names populate the table header and feature names populate the first column. Chromosome, position, ref, and alt populate the second, third, and fourth columns respectively. The table values are boolean: 1 for present mutation, 0 for missing mutation. Output: feature_and_mutation_by_sample.txt

all
Execute all output types

Known Issues
============
There is an issue when a variant file presents a contig that the pickled (archived) annotation does not have. The solution is to disable the archive feature by omitting the `-a` or `--archive_directory` option. This will permit the unknown contig in output, but all mutations on the unknown contig will be shown as having no feature. It may also be possible to make a new pickle file using an annotation GTF/GFF which does contain the contig in question.

    File "mucor/mucor.py", in parseVariantFiles
      resultSet = gas[ var.pos ]      # returns a set of zero to n IDs (e.g. gene symbols)
    File "_HTSeq.pyx", line 521, in HTSeq._HTSeq.GenomicArray.__getitem__ (src/_HTSeq.c:10700)
    KeyError: 

The VCF file needs to have columns #CHROM, POS â€¦ etc. The configuration script checks each VCF file for proper columns and will print a warning if any are missing or wrong. However, it does not halt execution and will include any malformed column VCF files and attempt to process them regardless. The main script may finish execution with the malformed VCF, but the output may be perturbed or useless.

    File "mucor/inputs.py", in parse_VarScan
        position = int(row[fieldId['POS']])
    KeyError: 'POS'

Users may not supply vcf files that have inconsistent 'effect' and/or 'functional consequence' for the same variant. Presumably, if the variant has the same location and reference and alternate allele, the effect and functional consequences would be predicted to be the same. The problem lies in collapsing mutations in the `variant_details` output type(s). This issue may arise when different platforms or pipelines are used for samples containing a common variant within a single run of VAS. If this is found to be more common than expected it will be worked around.

Bug Reports and Issue Tracking
==============================

Check the issue tracker at the bitbucket repository. Please report bugs and request new features through this interface for better tracking.

Contact
=======
james.blachly@osumc.edu

